presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduSplitSource.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kudu;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.spi.connector.ConnectorPartitionHandle;
+import io.prestosql.spi.connector.ConnectorSplitSource;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class KuduSplitSource
+        implements ConnectorSplitSource
+{
+    private final Map<Integer, KuduSplit> splitMap;
+
+    KuduSplitSource(List<KuduSplit> splits)
+    {
+        this.splitMap = new ConcurrentHashMap<>();
+        splits.forEach(split ->
+                splitMap.put(split.getBucketNumber(), split));
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+    {
+        KuduPartitionHandle kuduPartitionHandle = (KuduPartitionHandle) partitionHandle;
+        KuduSplit kuduSplit = splitMap.remove(kuduPartitionHandle.getBucket());
```

Comment:
Instead of removing the bucket, can we have a counter which gets incremented so that it would be helpful in `isFInished` ?

Reply:
Yes, I agree with you!  :)